File: src/preprocess.py
from pyspark.ml.feature import VectorAssembler, StandardScaler
from typing import List
from .logger import Logger
from .db_connector import MySQLConnector

class Preprocessor:
    def __init__(self, logger) -> None:
        self.features_name = 'features'
        self.scaled_features_name = 'scaled_features'
        self.separator = '\t'
        self.log = logger

    @Logger.cls_se_log("Загрузка датасета из БД")
    def load(self, db_connector: MySQLConnector, table: str, base_features: List[str]):
        dataset = db_connector.read(table)

        vector_assembler = VectorAssembler(
            inputCols=base_features,
            outputCol=self.features_name,
            handleInvalid='skip',
        )

        vectorized_data = vector_assembler.transform(dataset)
        return vectorized_data
    
    @Logger.cls_se_log("Применение к экземпларам датасета StandardScaler-трансформации")
    def apply_scale(self, vectorized_data):
        scaler = StandardScaler(
            inputCol=self.features_name,
            outputCol=self.scaled_features_name
        )

        scaler_model = scaler.fit(vectorized_data)
        scaled_data = scaler_model.transform(vectorized_data)

        return scaled_data

File: src/datamart/preprocess.scala

import org.apache.spark.ml.feature.{StandardScaler, VectorAssembler}
import org.apache.spark.sql.DataFrame
import com.typesafe.scalalogging.Logger


object Preprocessor {

  private val logger = Logger("Logger")

  def load(df: DataFrame): DataFrame = {
    logger.info("Старт загрузки датасета из БД...")
    val outputCol = "features"
    val inputCols = "code" :: "created_t" :: "last_modified_t" :: "completeness"

    val vector_assembler = new VectorAssembler()
      .setInputCols(df.columns)
      .setOutputCol(outputCol)
      .setHandleInvalid("skip")
    val result = vector_assembler.transform(df)
    logger.info("Датасет загружен!")
    result
  }

  def apply_scale(df: DataFrame): DataFrame = {
    logger.info("Старт применения к экземпларам датасета StandardScaler-трансформации")
    val scaler = new StandardScaler()
      .setInputCol("features")
      .setOutputCol("scaled_features")
    val scalerModel = scaler.fit(df)
    val result = scalerModel.transform(df)
    logger.info("Трансформация применена!")
    result
  }
}